### ISSUE TITLE
Super Metroid Z-Factor (romhack) broken scrolling

### ISSUE THREAD

**Issue by testyourmine** [byuu/bsnes#294]
The romhack Super Metroid Z-Factor is rather broken on bsnes v114.
[On bsnes](https://streamable.com/we3s2)
[How it looks working](https://www.youtube.com/watch?v=XS2Z9TkrBeg)
[settings.txt](https://github.com/byuu/bsnes/files/4176463/settings.txt)
It works as expected on snes9x 1.60 and Mesen-S latest nightly (and original hardware based off videos), but not bsnes or higan. I've tried fresh installing bsnes, tried the newest bsnes nightly, tried both fast and accurate ppu, and enabling the legacy settings, but nothing seems to work. I know it's not an official Nintendo game, but it's odd that it works on everything but bsnes and higan, and would like to play it on bsnes if it can be fixed as bsnes is my preferable snes emu of choice.

**Comment by byuu** [comment]
I apologize, but I don't support ROM hacks -- only test ROMs that demonstrate the issues in isolation so that I can fix them. I don't have the bandwidth or resources to troubleshoot what entire game ROM hacks are doing. I'll leave this report open for a while in the sincerest hopes that someone else is willing to lend a hand and look into it, but if it gets no traction, I'll be forced to close this as wontfix.

**Comment by Max833** [comment]
> It works as expected on snes9x 1.60 and Mesen-S latest nightly (and original hardware based off videos), but not bsnes or higan.

Just tested this. Super Metroid Z-Factor v1.3 is working on real hardware and higan 106.

**Comment by testyourmine** [comment]
I tested versions Super Metroid Z-Factor versions 1.0 through 1.2 and they all work, but for some reason v1.3 doesn't work. I can also confirm that it works on higan v106, but not v107. So somehow v1.3 did something that bsnes and higan doesn't like.

**Comment by byuu** [comment]
> For whatever it's worth... I just tested this. Super Metroid Z-Factor v1.3 is working on real hardware and higan 106.

Okay, so if someone wants to be super helpful, can they go through each commit and find what is causing the issue? ^-^;

**Comment by testyourmine** [comment]
Is there a guide on how to build bsnes? Because if so, I will go through each commit and see if I can find the version which broke it.

**Comment by Screwtapello** [comment]
The higan v106 instructions for [Windows](https://higan.readthedocs.io/en/v106/install/windows/#compiling-from-source-on-windows) and [Linux](https://higan.readthedocs.io/en/v106/install/linux/#installing-a-compiled-build-on-linux) shouldn't be too far out of date, although you might want to install the latest MSYS2 instead of trying to find TDM-GCC.

If you're familiar with git, I definitely recommend using the "git bisect" functionality.

**Comment by byuu** [comment]
It's a shame we didn't have cirrus-ci up sooner ^-^;
Though I think it drops binaries after a month anyway ... so maybe that wouldn't help.

**Comment by RokkumanX** [comment]
Although [THIS](https://gitlab.com/higan/higan/-/jobs) only goes up to v106r180, the downloads is still functional!

Better than nothing if I may say so myself...

**Comment by RokkumanX** [comment]
@testyourmine, @byuu, @Screwtapello 

Although [THIS](https://gitlab.com/higan/higan/-/jobs) only goes up to v106r180, the downloads is still functional!

Better than nothing if I may say so myself...

**Comment by testyourmine** [comment]
I was actually able to almost pinpoint the version that broke it on Wednesday. Higan v106 r8 is the last version of higan able to run Z-Factor correctly. Higan v106 r12 is the first playable version that breaks Z-Factor, similar to how it is now. We run into a problem in that r9, r10, and r11 all can't boot Z-Factor, due to the restructuring of icarus syntax and such. That's about as far as I have gotten, but perhaps it's possible for us to edit the code of those versions to allow Z-Factor to run and see which one really broke it, but it has yet to be done (it's only been a few days tho, we'll see what the future holds).

**Comment by jbo-85** [comment]
The size in the generated manifest by icarus is wrong, it's supposed to be 0x328000 and not 0x300000

Try this manifest in v106r12
```
game
  sha256:   129fcd596adb444f821f0d12b04b26a04002f97a30839339bebcb222eb2d54db
  label:    M3 Z-Factor Hack
  name:     sm
  region:   JPN
  revision: 1.0
  board:    LOROMEX-RAM
    memory
      type: ROM
      size: 0x328000
      category: Program
    memory
      type: RAM
        battery
      size: 0x2000
      category: Save
```

**Comment by testyourmine** [comment]
Yeah, I figured that out earlier today while comparing r8 and r12 manifests. So apparently the story is that most snes games are multiples of .5 for its size, .5, 1, 1.5, 2, etc MB of data. Z Factor is 3.15mb. The way bsnes/higan works is that unlike most emulators, which would just mirror the snes address or smth and add a bunch of data so thats its a multiple of .5, bsnes actually assumes its garbage data and will cut off all the data after the 3mb. I'm sure someone could explain it better, but that's my understanding from what someone said.
So expanding the rom to 3.5mb actually fixes the game since it doesn't cut off any data, but hopefully Near does something so that this doesn't happen or is an option or something similar.

**Comment by jbo-85** [comment]
```
auto SuperFamicom::romSize() const -> uint {
  //subtract appended firmware size, if firmware is present
  if((size() &  0x7fff) ==   0x100) return size() -   0x100;
  if((size() &  0x7fff) ==   0xc00) return size() -   0xc00;
  if((size() &  0x7fff) ==  0x2000) return size() -  0x2000;
  if((size() &  0xffff) ==  0xd000) return size() -  0xd000;
  if((size() & 0x3ffff) == 0x28000) return size() - 0x28000; // <-- That's the problem
  return size();
}
```

**Comment by Screwtapello** [comment]
OK, so Super Metroid is six 0.5MB blocks, and the Z-Factor romhack adds 160KB of extra data which doesn't quite fill up the seventh 0.5MB block.

Hayazashi Nidan Morita Shougi 2 is one 0.5MB block, and the extra AI firmware adds 160KB of extra data.

higan is designed to allow *any* game of any size to include the same co-processor that HNMS2 uses, and therefore allows any game to include a custom 160KB firmware blob, which it removes before starting emulation. bsnes and byuu inherit higan's behaviour, because they both want to enable higan-tier accuracy.

So bsnes looks at the Z-Factor ROM, sees that it is *exactly* 160KB more data than a multiple of 0.5MB, and slices off the "firmware blob" before loading the rest of the game. If Z-Factor used one byte more or less, it would be fine.

**Comment by devinacker** [comment]
Maybe the ideal solution is to just avoid truncating the ROM size unless the game actually requires an additional firmware ROM in the first place.

**Comment by byuu** [comment]
Correct. I'll see what can be done. I never expected someone to make a 3.15MB ROM. That is .......................... yeah.

**Comment by FitzRoyX** [comment]
That is one strange backfire of blobbing data together. Chasing a ghost bug for days and it ends up being rom expansion tripping the firmware detection code... wow.

**Comment by byuu** [comment]
This is fixed, thank you devinacker and testyourmine.
